### PR TITLE
FIR IDE: refactor KotlinFindUsagesSupport...

### PR DESCRIPTION
**Pull request by tgeng**
so that analyse can be done once for the entire container element rather
than on each element individually.

### CODE REVIEW

File: plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupportFirImpl.kt [hunk 1]
```diff
@@ -9,22 +9,26 @@ import com.intellij.psi.PsiElement
...
 
 class KotlinFindUsagesSupportFirImpl : KotlinFindUsagesSupport {
-    override fun isCallReceiverRefersToCompanionObject(element: KtElement, companionObject: KtObjectDeclaration): Boolean {
-        return false
+    override fun <T> analyseIfCallReceiverRefersToCompanionObject(
+        analyseScope: KtElement,
+        companionObject: KtObjectDeclaration,
+        block: (isCallReceiverRefersToCompanionObject: (KtElement) -> Boolean) -> T
+    ): T {
+        return block { false }
```
Comment (jsjeon) [inline plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupportFirImpl.kt:31]: Somehow we need to clarify/specify that some of `block` will be executed while some are not, depending on what `isCallReceiverRefersToCompanionObject` returns? For example, this `block { false }` looks like `block` will _not_ be executed no matter what. :\

Comment (tgeng) [reply plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupportFirImpl.kt:31]: I am not sure I understand what you would like me to do here. This code is incomplete for now. Let me add some TODO to make this clear.

File: plugins/kotlin/frontend-independent/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupport.kt [hunk 2]
```diff
@@ -29,14 +34,18 @@ interface KotlinFindUsagesSupport {
...
-        fun getSuperMethods(declaration: KtDeclaration, ignore: Collection<PsiElement>?) : List<PsiElement> =
+        fun getSuperMethods(declaration: KtDeclaration, ignore: Collection<PsiElement>?): List<PsiElement> =
             getInstance(declaration.project).getSuperMethods(declaration, ignore)
 
         fun sourcesAndLibraries(delegate: GlobalSearchScope, project: Project): GlobalSearchScope =
             getInstance(project).sourcesAndLibraries(delegate, project)
     }
 
-    fun isCallReceiverRefersToCompanionObject(element: KtElement, companionObject: KtObjectDeclaration): Boolean
+    fun <T> analyseIfCallReceiverRefersToCompanionObject(
```
Comment (darthorimar) [inline plugins/kotlin/frontend-independent/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupport.kt:44]: Name and semantic of the function is not very clear. Let's just move and make abstract to the `KotlinFindUsagesSupport` the whole `KotlinFindClassUsagesHandler.processCompanionObjectInternalReferences`

**Review by darthorimar — CHANGES_REQUESTED** [review]

File: plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupportFirImpl.kt (hunk 1 above)
Comment (tgeng) [reply plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/findUsages/KotlinFindUsagesSupportFirImpl.kt]: Per Ilya's suggestion, I have removed this and make the entire `processCompanionObjectInternalReferences` abstract.

**Comment by darthorimar** [comment]
Pushed, thank you!

**Pull request closed without merging**